### PR TITLE
Add jenkins integration

### PR DESCRIPTION
**Pull request by phalbert**
# Description

Added jenkins exporter for jobs and builds

## Type of change

Please leave one option from the following and delete the rest:
- [ ] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: integrations/jenkins/core/types.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    builtOn: str
+    changeSet: Optional[dict]
+    culprits: List[Optional[dict]]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(**kwargs)
+        self.timestamp = self.time_as_datetime
+
+    @property
+    def time_as_datetime(self):
```
Comment (PeyGis) [inline integrations/jenkins/core/types.py:67]: Why don't you use the convert_timestamp_to_utc_dt function in the utils.py?

File: integrations/jenkins/main.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    with logger.contextualize(event_id=event.id, event_state=event.type):
+        logger.info(f"{event.kind}: Received {event.dataType} event {event.id} | {event.type}")
+
+        if event.type in ["run.initialize", "run.started"]:
+            return {"ok": True}
+
+        event_record = event.dict(by_alias=True)
+        logger.info(event_record)
+
+        print(event_record)
```
Comment (PeyGis) [inline integrations/jenkins/main.py:56]: remove this

File: integrations/jenkins/main.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        await ocean.register_raw(event.kind, [event_record])
+        return {"ok": True}
+
+
+# Optional
```
Comment (PeyGis) [inline integrations/jenkins/main.py:61]: Remove this commented codes

**Review by PeyGis — COMMENTED** [review]
Looks good overall. Go over my comments and apply the fixes. Also, you need to add the changelog


towncrier create --content "add description here" version_number.type.md
repeat for each new feature
commit the change log files
 towncrier build --yes --version <version number>
 
 where type can be feature, improvement, bugfixes etc

**Comment by phalbert** [comment]
@PeyGis here are the changes.

File: integrations/jenkins/.port/spec.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+type: jenkins
+description: Jenkins integration for Port Ocean
+icon: Jenkins
+features:
+  - type: exporter
+    section: Git Providers
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/spec.yaml:6]: Ill need to tell you which section it should be

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+createMissingRelatedEntities: true
+deleteDependentEntities: true
+resources:
+  - kind: job
+    selector:
+      query:  "true"
+    port:
+      entity:
+        mappings:
+          identifier: 'if .url | contains("://") then (.url | split("://")[1] | sub("^.*?/"; "")) else .url end | sub("%20"; "-"; "g") | sub("/"; "-"; "g") | .[:-1]'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:10]: is there any other way to get the id for the job & build?

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+          title: '(.data.displayName // .displayName)'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:11]: why the difference?

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+          identifier: '(.data.fullDisplayName // .fullDisplayName) | sub(" "; "-"; "g") | sub("#"; ""; "g")'
+          title: '.data.displayName // .displayName'
+          blueprint: '"build"'
+          properties:
+            buildStatus: '.data.result // .result'
+            buildUrl: '.url | if test("://") then sub("^[^/]+//[^/]+"; "") else . end'
+            buildDuration: '.data.duration // .duration'
+            timestamp: '(.time // .timestamp) as $input | if $input | type == "number" then ($input / 1000 | todate) else $input end'
+          relations:
+            job: '(.source // .url) as $input | if $input | contains("://") then ($input | split("://")[1] | sub("^.*?/"; "")) else $input end | tostring | sub("%20"; "-"; "g") | sub("/"; "-"; "g") | .[:-1] | sub("-[0-9]+$"; "")'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:33]: The build dont hold the job id?

if not then lets talk about how to ingest custom information to the data

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+      entity:
+        mappings:
+          identifier: '(.data.fullDisplayName // .fullDisplayName) | sub(" "; "-"; "g") | sub("#"; ""; "g")'
+          title: '.data.displayName // .displayName'
+          blueprint: '"build"'
+          properties:
+            buildStatus: '.data.result // .result'
+            buildUrl: '.url | if test("://") then sub("^[^/]+//[^/]+"; "") else . end'
+            buildDuration: '.data.duration // .duration'
+            timestamp: '(.time // .timestamp) as $input | if $input | type == "number" then ($input / 1000 | todate) else $input end'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:31]: why are there 2 states?

File: integrations/jenkins/README.md
Comment (yairsimantov20) [inline integrations/jenkins/README.md:1]: Dont forget to update any changes after we finish working on this

File: integrations/jenkins/config.yaml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+# This is an example configuration file for the integration service.
+# Please copy this file to config.yaml file in the integration folder and edit it to your needs.
+
```
Comment (yairsimantov20) [inline integrations/jenkins/config.yaml:3]: it should create default resources

File: integrations/jenkins/config.yaml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+port:
+  clientId: "{{ from env PORT_CLIENT_ID }}" # Can be loaded via environment variable: PORT_CLIENT_ID
+  clientSecret: "{{ from env PORT_CLIENT_SECRET }}" # Can be loaded via environment variable: PORT_CLIENT_SECRET
+# The event listener to use for the integration service.
+eventListener:
+  type: POLLING
+integration:
+  # The identifier of this integration instance.
+  identifier: "my_jenkins_integration"
```
Comment (yairsimantov20) [inline integrations/jenkins/config.yaml:12]:
```suggestion
  identifier: "my-jenkins-integration"
```

File: integrations/jenkins/core/client.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+
+class JenkinsClient:
+    def __init__(
+        self, jenkins_base_url: str, jenkins_user: str, jenkins_password: str
+    ) -> None:
+        self.jenkins_base_url = jenkins_base_url
+
+        auth = (jenkins_user, jenkins_password)
+        self.client = httpx.AsyncClient(auth=auth)
```
Comment (yairsimantov20) [inline integrations/jenkins/core/client.py:16]: lets switch to the global http client in the latest ocean version

File: integrations/jenkins/core/client.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+from loguru import logger
+
+
+class JenkinsClient:
+    def __init__(
+        self, jenkins_base_url: str, jenkins_user: str, jenkins_password: str
+    ) -> None:
+        self.jenkins_base_url = jenkins_base_url
+
+        auth = (jenkins_user, jenkins_password)
```
Comment (yairsimantov20) [inline integrations/jenkins/core/client.py:15]: can we use a token instead?

File: integrations/jenkins/core/types.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    BUILD = "build"
+
+    @staticmethod
+    def get_object_kind_for_event(obj_type: str):
+        if obj_type.startswith("item"):
+            return ObjectKind.JOB
+        elif obj_type.startswith("run"):
+            return ObjectKind.BUILD
+        else:
+            return None
```
Comment (yairsimantov20) [inline integrations/jenkins/core/types.py:15]: This can be done in the route
no need for the extra function lets keep it simple

File: integrations/jenkins/core/types.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
+from enum import StrEnum
+
+
+class ObjectKind(StrEnum):
```
Comment (yairsimantov20) [inline integrations/jenkins/core/types.py:4]: hold this enum in the main file

File: integrations/jenkins/main.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+@ocean.on_resync(ObjectKind.BUILD)
+async def on_resync_builds(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    jenkins_client = init_client()
+
+    async for jobs in jenkins_client.get_jobs():
+        logger.info(f"Received ${len(jobs)} jobs")
+        for job in jobs:
+            async for builds in jenkins_client.get_builds(job["name"]):
+                yield builds
```
Comment (yairsimantov20) [inline integrations/jenkins/main.py:35]: I have an idea for implementation

File: integrations/jenkins/main.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+
+@ocean.router.post("/events")
+async def handle_events(event: dict[str, Any]) -> dict[str, bool]:
+    logger.info(
+        f'Received {event["dataType"]} event {event["id"]} | {event["type"]}'
+    )
+
+    await ocean.register_raw(ObjectKind.get_object_kind_for_event(event["type"]), [event])
+    return {"ok": True}
```
Comment (yairsimantov20) [inline integrations/jenkins/main.py:45]: How did you implement it?
can we create webhooks for them?

**Comment by yairsimantov20** [comment]
Update to the latest ocean version

**Comment by yairsimantov20** [comment]
suggested implementation


```
@ocean.on_resync()
async def on_resync(kind: str) -> None:
    async for _jobs in get_all_jobs_and_builds(kind):
        if kind == "Job":
            yield _jobs
        elif kind == "Build":
            yield [build for job in _jobs for build in job["builds"]]

File: integrations/jenkins/main.py (hunk 16 above)
Comment (phalbert) [reply integrations/jenkins/main.py:45]: we cannot create a webhook using API. A user has to manually set integration hook in Jenkins

File: integrations/jenkins/core/client.py (hunk 12 above)
Comment (phalbert) [reply integrations/jenkins/core/client.py:15]: either a token or password works. i will change the variable name to token though

File: integrations/jenkins/.port/resources/port-app-config.yaml (hunk 8 above)
Comment (phalbert) [reply integrations/jenkins/.port/resources/port-app-config.yaml:31]: this was because of different json schemas from the jenkins api and the jenkins events.But we are going to use only the api schema now

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
comments

File: integrations/jenkins/.port/resources/port-app-config.yaml (hunk 5 above)
Comment (phalbert) [reply integrations/jenkins/.port/resources/port-app-config.yaml:10]: The build and job objects both contain the URL as their primary identifier. This ensures the integrity of the mapping between builds and jobs, as we can always rely on the URLs to maintain a consistent reference. For example, a job named "Project Name" with builds 34 and 35 would have the following URLs:

Job: `https://jenkin_host/job/Project%20Name`
Builds:

- `https://jenkin_host/job/Project%20Name/34`
- `https://jenkin_host/job/Project%20Name/35`

While using other fields like name or displayName as identifiers could be an option, it could complicate the process of determining the job ID from a build URL. We would need to parse the URL and extract the relevant information to match it with the job ID we expect.

File: integrations/jenkins/main.py [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+                yield builds
+
+
+@ocean.router.post("/events")
+async def handle_events(event: dict[str, Any]) -> dict[str, bool]:
+    logger.info(
+        f'Received {event["dataType"]} event {event["id"]} | {event["type"]}'
+    )
+
+    await ocean.register_raw(ObjectKind.get_object_kind_for_event(event["type"]), [event])
```
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/jenkins/main.py]: Oh, this is an old file. I had already made the change.

File: integrations/jenkins/CHANGELOG.md [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+- Use the global httpx client (0.1.1)
+- Get jobs and builds at once in resync (0.1.1)
+- Simplify code structure and organisation (0.1.1)
+- Add caching to resync (0.1.1)
+- Bumped ocean version to ^0.4.7 (0.1.1)
+
+### Improved Documentation
+
+- Update examples for port-app-config (0.1.1)
+
```
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/jenkins/CHANGELOG.md]: removed for now

File: integrations/jenkins/config.yaml (hunk 9 above)
Comment (phalbert) [reply integrations/jenkins/config.yaml]: This works now

File: integrations/jenkins/core/client.py [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+        page = 0
+        logger.info("Getting jobs from Jenkins")
+
+        try:
+            while True:
+                start_idx = page_size * page
+                end_idx = start_idx + page_size
+
+                params = {
+                    "tree": f"jobs[name,url,description,displayName,fullDisplayName,fullName]{{{start_idx},{end_idx}}}"
```
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/jenkins/core/client.py]: This is now added in the current file

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 20]
```diff
@@ -0,0 +1,35 @@
+createMissingRelatedEntities: true
+deleteDependentEntities: true
+resources:
+  - kind: jenkinsJob
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:4]: The kinds should not be prefixed

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    selector:
+      query:  "true"
+    port:
+      entity:
+        mappings:
+          identifier: 'if .url | contains("://") then (.url | split("://")[1] | sub("^.*?/"; "")) else .url end | sub("%20"; "-"; "g") | sub("/"; "-"; "g") | .[:-1]'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:10]: I prefer sticking with ids over the url
the url might change between jenkins versions

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+          title: .displayName
+          blueprint: '"jenkinsJob"'
+          properties:
+            jobName: .fullName
+            url: '.url | if test("://") then sub("^[^/]+//[^/]+"; "") else . end'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:15]: what is the purpose of this jq?

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+            jobStatus: '.color as $input | if $input == "notbuilt" then "created" else "updated" end'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:16]: Why are you using color
the job dont have status?

Also this jq can be simplefied

```suggestion
            jobStatus: 'if .color == "notbuilt" then "created" else "updated" end'
```

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+            timestamp: .time
+          relations:
+            jobs: '.jobs | map(if .url | contains("://") then (.url | split("://")[1] | sub("^.*?/"; "")) else .url end | sub("%20"; "-"; "g") | sub("/"; "-"; "g") | .[:-1])'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:19]: can we again use ids?

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+  - kind: jenkinsBuild
+    selector:
+      query: "true"
+    port:
+      entity:
+        mappings:
+          identifier: '.fullDisplayName | sub(" "; "-"; "g") | sub("#"; ""; "g") | sub("»"; ""; "g") '
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:26]: please use the identifiers display names might change

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+          title: .displayName
+          blueprint: '"jenkinsBuild"'
+          properties:
+            buildStatus: .result
+            buildUrl: '.url | if test("://") then sub("^[^/]+//[^/]+"; "") else . end'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:31]: why is this needed

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+            buildDuration: .duration
+            timestamp: '.timestamp as $input | if $input | type == "number" then ($input / 1000 | todate) else $input end'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:33]: can this jq be simplified?

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+          relations:
+            job: '.url as $input | if $input | contains("://") then ($input | split("://")[1] | sub("^.*?/"; "")) else $input end | tostring | sub("%20"; "-"; "g") | sub("/"; "-"; "g") | .[:-1] | sub("-[0-9]+$"; "")'
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:35]: please use ids

File: integrations/jenkins/.port/resources/blueprints.json [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+                    "title": "Build Duration",
+                    "description": "Duration of the build"
+                }
+            },
+            "required": []
+        },
+        "mirrorProperties": {},
+        "calculationProperties": {},
+        "relations": {
+            "job": {
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/blueprints.json:86]: use better title and identifiers to explain what this is

File: integrations/jenkins/main.py [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+from loguru import logger
+from enum import StrEnum
+
+from client import JenkinsClient
+from port_ocean.context.ocean import ocean
+
+
+class ObjectKind(StrEnum):
+    JOB = "jenkinsJob"
+    BUILD = "jenkinsBuild"
```
Comment (yairsimantov20) [inline integrations/jenkins/main.py:11]: please leave the kinds without the prefix

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
Left more comments

**Comment by yairsimantov20** [comment]
upgrade ocean version

File: integrations/jenkins/.port/resources/port-app-config.yaml (hunk 22 above)
Comment (phalbert) [reply integrations/jenkins/.port/resources/port-app-config.yaml:15]: Oh, this is a left-over from when I was trying to handle the two json schemas of `events` and `api`. I will simplify it.

File: integrations/jenkins/.port/resources/port-app-config.yaml (hunk 23 above)
Comment (phalbert) [reply integrations/jenkins/.port/resources/port-app-config.yaml:16]: The `color` field is actually the job status. It shows the last successful build status. But I will simplify the JQ

File: integrations/jenkins/.port/resources/port-app-config.yaml (hunk 27 above)
Comment (phalbert) [reply integrations/jenkins/.port/resources/port-app-config.yaml:33]: yes, it can be. I will simplify it

File: integrations/jenkins/.port/resources/port-app-config.yaml (hunk 26 above)
Comment (phalbert) [reply integrations/jenkins/.port/resources/port-app-config.yaml:31]: I have simplified this

File: integrations/jenkins/.port/resources/port-app-config.yaml (hunk 25 above)
Comment (phalbert) [reply integrations/jenkins/.port/resources/port-app-config.yaml]: so we now have a `__jobName` field

File: integrations/jenkins/.port/resources/port-app-config.yaml (hunk 28 above)
Comment (phalbert) [reply integrations/jenkins/.port/resources/port-app-config.yaml]: using `.name` for each record.

File: integrations/jenkins/.port/resources/blueprints.json (hunk 29 above)
Comment (phalbert) [reply integrations/jenkins/.port/resources/blueprints.json]: changed this to `parentJob`

File: integrations/jenkins/main.py (hunk 30 above)
Comment (phalbert) [reply integrations/jenkins/main.py]: changed them back and removed the prefix

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 31]
```diff
@@ -16,14 +16,14 @@ resources:
             jobStatus: '.color as $c | {"notbuilt": "created", "blue": "passing", "red": "failing"}[$c // "updated"]'
             timestamp: .time
           relations:
-            jobs: '.jobs | map(if .url | contains("://") then (.url | split("://")[1] | sub("^.*?/"; "")) else .url end | sub("%20"; "-"; "g") | sub("/"; "-"; "g") | .[:-1])'
+            jobs: '. as $body | $body.jobs | map($body.name + " " + .name  | sub(" "; "-"; "g")) '
```
Comment (phalbert) [inline integrations/jenkins/.port/resources/port-app-config.yaml:19]: @yairsimantov20, this is so because if two jobs "JobA" and "JobB" both have `main` branches, then the identifiers become "JobA-main" and "JobB-main"

File: integrations/jenkins/main.py [hunk 32]
```diff
@@ -49,6 +53,9 @@ async def handle_events(event: dict[str, Any]) -> dict[str, bool]:
 
     if kind:
         resource = await jenkins_client.get_single_resource(event["url"])
+        if kind == ObjectKind.BUILD:
+            # the only way to get the job name is from the build url
+            resource = {**resource, "__jobName": parse_job_name(resource["url"])}
```
Comment (phalbert) [inline integrations/jenkins/main.py:58]: We still have to parse the URL in order to get the job name for a build  webhook event.

Comment (phalbert) [reply integrations/jenkins/main.py:58]: Would you rather we use the `fullDisplayName`. I am more wary of its change. I am more certain of the `/job/<jobName>/` since the <jobName> is from the "name" field of the job json data

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]

**Comment by phalbert** [comment]
### [Only save buildable jobs. And save their parents in a nested key.](https://github.com/port-labs/ocean/pull/233#:~:text=Recursively%20get%20buildable%20jobs)

How a job under a hierarchy of jobs looks like. This is the `main` whose parent job is `limbo-core` whose parent is `Limbopay`

Screenshot 2023-12-27 at 19 31 28


How the `__parentJob` looks like in the UI.

Screenshot 2023-12-27 at 19 29 57

File: integrations/jenkins/client.py [hunk 33]
```diff
@@ -0,0 +1,103 @@
...
+            yield all_jobs
+            event.attributes[cache_key] = all_jobs
+        except Exception as e:
+            logger.error(f"Unexpected error occurred: {e}")
+            raise
+
+    async def fetch_jobs_and_builds_from_api(
+        self, parent_job: Optional[dict[str, Any]] = None
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        page_size = 5
```
Comment (yairsimantov20) [inline integrations/jenkins/client.py:44]: why is page size 5?

File: integrations/jenkins/client.py [hunk 34]
```diff
@@ -0,0 +1,103 @@
...
+            async for jobs in self.fetch_jobs_and_builds_from_api():
+                all_jobs.extend(jobs)
+
+            yield all_jobs
+            event.attributes[cache_key] = all_jobs
+        except Exception as e:
+            logger.error(f"Unexpected error occurred: {e}")
+            raise
+
+    async def fetch_jobs_and_builds_from_api(
```
Comment (yairsimantov20) [inline integrations/jenkins/client.py:41]: this function is way overloaded 
separate this pls
at the moment it is unreadable

you have recursions and loops and if checks way 2 much logic

File: integrations/jenkins/client.py [hunk 35]
```diff
@@ -0,0 +1,103 @@
...
+                "tree": f"jobs[name,url,displayName,fullName,color,jobs[name,color,fullName,displayName,url],"
+                f"builds[id,number,url,result,duration,timestamp,displayName,fullDisplayName]"
+                f"{{0,50}}]{{{start_idx},{end_idx}}}"
+            }
+
+            if parent_job:
+                job_path = urlparse(parent_job["url"]).path
+                base_url = f"{self.jenkins_base_url}{urlparse(job_path).path}"
+            else:
+                base_url = self.jenkins_base_url
```
Comment (yairsimantov20) [inline integrations/jenkins/client.py:62]: why is this needed?
dont you get all the jobs when searching for jobs?

File: integrations/jenkins/client.py [hunk 36]
```diff
@@ -0,0 +1,103 @@
...
+
+        try:
+            async for jobs in self.fetch_jobs_and_builds_from_api():
+                all_jobs.extend(jobs)
+
+            yield all_jobs
+            event.attributes[cache_key] = all_jobs
+        except Exception as e:
+            logger.error(f"Unexpected error occurred: {e}")
+            raise
```
Comment (yairsimantov20) [inline integrations/jenkins/client.py:39]: why the broad exception handling?

File: integrations/jenkins/client.py [hunk 37]
```diff
@@ -0,0 +1,103 @@
...
+
+class JenkinsClient:
+    def __init__(
+        self, jenkins_base_url: str, jenkins_user: str, jenkins_token: str
+    ) -> None:
+        self.jenkins_base_url = jenkins_base_url
+        self.client = http_async_client
+        self.client.auth = httpx.BasicAuth(jenkins_user, jenkins_token)
+
+    async def get_all_jobs_and_builds(
```
Comment (yairsimantov20) [inline integrations/jenkins/client.py:19]: why do u have 2 functions with the same purpose?

File: integrations/jenkins/client.py [hunk 38]
```diff
@@ -0,0 +1,103 @@
...
+            job_response = await self.client.get(f"{base_url}/api/json", params=params)
+            job_response.raise_for_status()
+            jobs = job_response.json()["jobs"]
+
+            if not jobs:
+                break
+
+            for job in jobs:
+                if parent_job:
+                    parent_job["__jobsCount"] = len(parent_job["jobs"])
```
Comment (yairsimantov20) [inline integrations/jenkins/client.py:73]: why do you need it?

File: integrations/jenkins/client.py [hunk 39]
```diff
@@ -0,0 +1,103 @@
...
+                    job["__parentJob"] = {
+                        key: value
+                        for key, value in parent_job.items()
+                        if parent_job and key != "jobs"
+                    }
+
+                if "builds" not in job:
+                    # Recursively fetch child jobs only if no builds key
```
Comment (yairsimantov20) [inline integrations/jenkins/client.py:81]: wouldnt we get the jobs from the regular jobs route?

File: integrations/jenkins/client.py [hunk 40]
```diff
@@ -0,0 +1,103 @@
...
+                base_url = f"{self.jenkins_base_url}{urlparse(job_path).path}"
+            else:
+                base_url = self.jenkins_base_url
+
+            job_response = await self.client.get(f"{base_url}/api/json", params=params)
+            job_response.raise_for_status()
+            jobs = job_response.json()["jobs"]
+
+            if not jobs:
+                break
```
Comment (yairsimantov20) [inline integrations/jenkins/client.py:69]: why are u handling it here? arent you handling it in the end?

File: integrations/jenkins/.port/resources/port-app-config.yaml [hunk 41]
```diff
@@ -0,0 +1,43 @@
...
+    selector:
+      query: "true"
+    port:
+      entity:
+        mappings:
+          identifier: |
+            def parseURL:
+                  capture("^((?<scheme>[^:/?#]+):)?(//(?<authority>(?<domain>[^/?#:]*)(:(?<port>[0-9]*))?))?((?<path>[^?#]*))?(\\?(?<query>([^#]*)))?(#(?<fragment>(.*)))?");
+            
+            .url | parseURL | .path | gsub("%20"; "")  | gsub("/job/"; "/") | ascii_downcase | gsub("/"; "-") | gsub("^[/\\-]+|[/\\-]+$"; "")
```
Comment (yairsimantov20) [inline integrations/jenkins/.port/resources/port-app-config.yaml:29]: This just got the mapping way way way more complicated
if you can do something with jq dont force it please
clients need to understand it

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
the code is very messy right now
please go over my comments

File: integrations/jenkins/client.py (hunk 39 above)
Comment (phalbert) [reply integrations/jenkins/client.py:81]: We can get the jobs through the regular jobs route. The challenge, however, is that we don't know how deep the hierarchy would be. So we fetch the first two levels and then keep checking if a job doesn’t have a `builds` key in order to fetch the child jobs.

File: integrations/jenkins/client.py (hunk 38 above)
Comment (phalbert) [reply integrations/jenkins/client.py:73]: I thought it would be a useful addition to the job details shown in the UI under the parentJob json attribute.

File: integrations/jenkins/.port/resources/port-app-config.yaml (hunk 41 above)
Comment (phalbert) [reply integrations/jenkins/.port/resources/port-app-config.yaml:29]: I understand you on this. Please check out this slack [thread](https://getport.slack.com/archives/D0696BJPNK1/p1703595840746549?thread_ts=1703164476.063059&cid=D0696BJPNK1) where I had raised my concern over this issue. It seemed like we had agreed on using a custom field of `__jobName` but the last [feedback](https://getport.slack.com/archives/D0696BJPNK1/p1703164476063059) went in a different direction of going back to jq.

File: integrations/jenkins/client.py (hunk 35 above)
Comment (phalbert) [reply integrations/jenkins/client.py:62]: This becomes valuable in the recursion process to distinguish between making a top-level call to retrieve all jobs and making a call on a specific job route to obtain its child jobs.

File: integrations/jenkins/client.py (hunk 33 above)
Comment (phalbert) [reply integrations/jenkins/client.py:44]: I left this as I was testing. I will remove it.

File: integrations/jenkins/client.py (hunk 34 above)
Comment (phalbert) [reply integrations/jenkins/client.py:41]: I am going to simplify the code and make it easier.

But I found recursion to be the best way to handle this scenario. Did you have an alternative way in mind?

File: integrations/jenkins/client.py (hunk 36 above)
Comment (phalbert) [reply integrations/jenkins/client.py:39]: I will simplify this.

File: integrations/jenkins/client.py (hunk 37 above)
Comment (phalbert) [reply integrations/jenkins/client.py:19]: I can refactor and simplify them.

File: integrations/jenkins/client.py (hunk 40 above)
Comment (phalbert) [reply integrations/jenkins/client.py:69]: Hmm, I will take a look at it again.
